swarm/src/handler.rs

```diff
@@ -508,25 +496,56 @@ impl<TConnectionUpgrade, TOutboundOpenInfo, TCustom, TErr>
             ConnectionHandlerEvent::Close(val) => ConnectionHandlerEvent::Close(val),
         }
     }
+}
 
-    /// If this is a `Close` event, maps the content to something else.
-    pub fn map_close<F, I>(
-        self,
-        map: F,
-    ) -> ConnectionHandlerEvent<TConnectionUpgrade, TOutboundOpenInfo, TCustom, I>
-    where
-        F: FnOnce(TErr) -> I,
-    {
-        match self {
-            ConnectionHandlerEvent::OutboundSubstreamRequest { protocol } => {
-                ConnectionHandlerEvent::OutboundSubstreamRequest { protocol }
-            }
-            ConnectionHandlerEvent::Custom(val) => ConnectionHandlerEvent::Custom(val),
-            ConnectionHandlerEvent::Close(val) => ConnectionHandlerEvent::Close(map(val)),
+/// Encapsulates the reason for why a specific [`ConnectionHandler`] closed a connection.
+#[derive(Debug)]
+pub struct CloseReason {
+    protocol: &'static str,
```

Comment:
To me this is a hack. The compiler is already aware of the place where a `CloseReason` is instantiated. Passing an additional string seems error prone to me. Is there no way we can leverage the type system to make passing a string unnecessary?

Reply:
I agree that this is a hack and I would like to get rid of it. If we agree [above](https://github.com/libp2p/rust-libp2p/pull/3201#discussion_r1061617524) that we want _something_ like this, I am happy to invest more time into this detail to make that better.

Reply:
One alternative as discussed above is recommend users to implement their own error type if they care about identifying the origin of the connection closure. We can follow this as a best-practice in our codebase.

This is kind of equivalent to what we have now. A user could already specify a generic error like `std::io::Error` here and thus essentially drop the information of who closed the connection. This assumes they don't want to rely on the order of how the error is composed as part of `derive(NetworkBehaviour)`. I think this is a fair assumption to make given that one could change the order of fields in a `NetworkBehaviour` without necessarily generating a compile error.